### PR TITLE
filtering/reconnecting

### PR DESCRIPTION
**Pull request by brauliorivas**
BEGINRELEASENOTES
- Filtering/reconnecting particles is a critical task for dmX. So, in this PR, I'm implementing a basic filtering/reconnect function that given a function that determines if a particle should be filtered out, reconnects the particles again.

ENDRELEASENOTES

### CODE REVIEW

File: js/main.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ const mouseUp = function (event) {
   isDragging = false;
 
   // console.time("drawAll");
-  drawAll();
+  drawAll(parentLinks, childrenLinks, infoBoxes);
```
Comment (tmadlener) [inline js/main.js:53]: Stupid question from my side, where do these inputs come from?

Comment (brauliorivas) [reply js/main.js:53]: Don't worry, these inputs come from [main.js](https://github.com/key4hep/dmx/blob/main/js/main.js#L18-L20).

**Comment by brauliorivas** [comment]
Hey @kjvbrt and @tmadlener, could you check my filtering function? As previously discussed, an external function determines whether to filter or not a particle. And then reconnected by using the `reconnect` function in `js/menu/filter.js`. I have added a toggle for simplicity but it's temporary.

**Comment by tmadlener** [comment]
I am rebasing this to potentially pick up the preview functionality.

File: js/menu/filter.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        particles
+      );
+      drawAll(newParentLinks, newChildrenLinks, newParticles);
+    } else {
+      drawAll(parentLinks, childrenLinks, particles);
+    }
+  }
+}
+
+function reconnect(criteriaFunction, parentLinks, childrenLinks, particles) {
```
Comment (tmadlener) [inline js/menu/filter.js:24]: I am not entirely sure whether this is necessary yet, but just for the future; Can this be written in a way where there are no `parentLinks`? We will probably need that e.g. for the `ReconstructedParticle` and I am wondering whether we should already spend the time now to prepare that?

**Comment by brauliorivas** [comment]
Hey @tmadlener @kjvbrt, I started working on more advanced filtering for `MonteCarlo Particles` (as task for week 1 of the proposal). It's a bit buggy, but I want to share how is it going. Here is a small demo.

8eadf647-25e3-41ec-b2be-c97043fb2e95

**Comment by tmadlener** [comment]
Very nice. Since you have the panel now already. Can I ask for one specific filter function which will should be straight forward to implement (from a technical point of view), but will probably require a bit of "leg work" from your side.

The `edm4hep::MCParticle` has two bitfields: `generatorStatus` and `simulatorStatus`. For the `simulatorStatus` we provide a few bit values that are reserved for a specific meaning:

https://github.com/key4hep/EDM4hep/blob/997ab32b886899253c9bc61adea9a21b57bc5a21/edm4hep.yaml#L294-L301

And then there is also some convenience functionality to check wheter a specifc bit is set:

https://github.com/key4hep/EDM4hep/blob/997ab32b886899253c9bc61adea9a21b57bc5a21/edm4hep.yaml#L306-L321

Would it be possible to have a filter that allows us to toggle these to filter out MCParticles for which specific bits are set? I.e. Have one checkbox / toggle for each bit and combine them into a filter function? IIUC all the filter combiners, etc. are already in place and this would mainly boil down to writing a specific filter for checking the bits in the `simulatorStatus` bit field.

**Comment by brauliorivas** [comment]
I have it now. In theory it should work, but the particles from the `json` example don't seem to have a simulator status from the ones described, so it's a bit difficult to test. Here's an image of the filters now.
Selection_001

**Comment by tmadlener** [comment]
Rebasing this again to see if the preview fixes from #22 are picked up.

**Comment by tmadlener** [comment]
Preview is working :rocket: However, for me it doesn't show the filter panel. Do I have to enable them somehow?

For testing the filtering, it would be nice to have some unittests for that to see if things work as expected. In these unittests you can just create some dummy MCParticle hierarchy with the values that you want in the end. You could start from the existing json file and just use a few of them and just adapt them to enable testing. This should all be possible without ever having to go to the UI to display them (for testing).

**Comment by brauliorivas** [comment]
> Preview is working 🚀 However, for me it doesn't show the filter panel. ~Do I have to enable them somehow?~ It looks like the preview doesn't publish the status of the PR, but rather the one from the PR. I am not sure why, though.

Yeah, it's strange. I didn't expect this. I'll fix it.



> For testing the filtering, it would be nice to have some unittests for that to see if things work as expected. In these unittests you can just create some dummy MCParticle hierarchy with the values that you want in the end. You could start from the existing json file and just use a few of them and just adapt them to enable testing. This should all be possible without ever having to go to the UI to display them (for testing).

Sure! In that case, I will wait for this PR and #20 to be merged, to add testing, because the testing setup is in #20 and this #18 has the function to be tested. So I think that after merging both PRs to main, I should create a separate PR to test filtering, and other PR to fix again PR previews, what do you think?

**Comment by tmadlener** [comment]
I am quite happy with this. I think we can merge this, unless @kjvbrt has more comments.

File: js/menu/filter/reconnect.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import { Link } from "../../objects";
```
Comment (kjvbrt) [inline js/menu/filter/reconnect.js:1]: `import { Link } from "../../objects";` ->
`import { Link } from "../../objects.js";`

**Comment by kjvbrt** [comment]
Otherwise it runs nicely. For next PRs we can think of showing only sim/gen status numbers which are used in the file.

**Comment by kjvbrt** [comment]
I run it again and found that the filter is reset, when I move any of the elements.

**Pull request merged**